url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/util/RStar.java

```diff
@@ -0,0 +1,2459 @@
+								result.add(second.getEntry(j).getSon());
+							}
+							break;
+						default:
+							break;
+						}
+					}
+				}
+			} else {
+				// both inner nodes - check their entries and go recursivly down
+				if (!(this.isLeaf()) && !(second.isLeaf())) {
+					for (int i = 0; i < this.getUsed(); i++) {
+						for (int j = 0; j < second.getUsed(); j++) {
+							// check for intersection - if intersects perform
+							// intersection join on their son nodes
+							if (this.getEntry(i).getMBB().intersectsStrict(
+									second.getEntry(j).getMBB(), _sop))
+								result = ((Node) this.getEntry(i).getSon())
+										.join((Node) second.getEntry(j)
+												.getSon(), result, predicate);
+						}
+					}
+				} else { // one is leaf, the other inner node
+					if (this.isLeaf()) { // this is leaf
+						// check for all leaf entry objects of second inner node
+						// for PREDICATE
+						for (int i = 0; i < this.getUsed(); i++) {
+							Set retrieveSet = new HashSet();
+
+							switch (predicate) {
+							case INTERSECTS_STRICT_PREDICATE:
+
+								second.retrieve(this.getEntry(i).getMBB(),
+										INTERSECTS_STRICT_PREDICATE,
+										retrieveSet);
+								break;
+							case EQUALS_PREDICATE:
+
+								second.retrieve(this.getEntry(i).getMBB(),
+										EQUALS_PREDICATE, retrieveSet);
+								break;
+							default:
+								break;
+							}
+							for (Iterator it = retrieveSet.iterator(); it
+									.hasNext();) {
+								result.add(this.getEntry(i).getSon());
+								result.add(it.next());
+							}
+						}
+					} else { // second is leaf
+						// check for all leaf entry objects of this inner node
+						// for PREDICATE
+						for (int i = 0; i < second.getUsed(); i++) {
+							Set retrieveSet = new HashSet();
+							switch (predicate) {
+							case INTERSECTS_STRICT_PREDICATE:
+								second.retrieve(this.getEntry(i).getMBB(),
+										INTERSECTS_STRICT_PREDICATE,
+										retrieveSet);
+								break;
+							case EQUALS_PREDICATE:
+								second.retrieve(this.getEntry(i).getMBB(),
+										EQUALS_PREDICATE, retrieveSet);
+								break;
+							default:
+								break;
+							}
+
+							for (Iterator it = retrieveSet.iterator(); it
+									.hasNext();) {
+								result.add(it.next());
+								result.add(second.getEntry(i).getSon());
+							}
+						}
+					}
+				}
+			}
+			return result;
+		}
+
+		/*
+		 * Retrieves entries based on the retrieve type.<br> retrieve type
+		 * possibilities:<br> 0 - intersects - result entries mbbs must
+		 * intersect with given MBB3D<br> 1 - contain - result entries mbbs must
+		 * contain the given MBB3D<br> 2 - inside - result entries mbbs have to
+		 * be inside the given MBB3D<br>
+		 * 
+		 * @param mbb test MBB3D
+		 * 
+		 * @param retrieveType int retrieve type
+		 * 
+		 * @param set result Set
+		 * 
+		 * @throws DB3DException.
+		 */
+		protected void retrieve(MBB3D mbb, int retrieveType, Set set) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.